crates/engine/tree/src/tree/root.rs

```diff
@@ -72,6 +84,78 @@ impl StdReceiverStream {
     }
 }
 
+/// Messages used internally by the state root task
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) enum StateRootMessage {
+    /// New state update from transaction execution
+    StateUpdate(EvmState),
+    /// Proof calculation completed for a specific state update
+    ProofCalculated {
+        /// The calculated proof
+        proof: MultiProof,
+        /// The index of this proof in the sequence of state updates
+        sequence_number: u64,
+    },
+    /// State root calculation completed
+    RootCalculated {
+        /// The calculated state root
+        root: B256,
+        /// The trie updates produced during calculation
+        updates: TrieUpdates,
+        /// Time taken to calculate the root
+        elapsed: Duration,
+    },
+}
+
+/// Handle to track proof calculation ordering
+#[derive(Debug, Default)]
+pub(crate) struct ProofSequencer {
+    /// The next expected proof sequence number
+    next_sequence: u64,
+    /// Buffer for out-of-order proofs
+    pending_proofs: BTreeMap<u64, MultiProof>,
+}
+
+impl ProofSequencer {
+    /// Creates a new proof sequencer
+    pub(crate) fn new() -> Self {
+        Self::default()
+    }
+
+    /// Gets the next sequence number and increments the counter
+    pub(crate) fn next_sequence(&mut self) -> u64 {
+        let seq = self.next_sequence;
+        self.next_sequence += 1;
+        seq
+    }
+
+    /// Adds a proof and returns all sequential proofs if we have a continuous sequence
+    pub(crate) fn add_proof(&mut self, sequence: u64, proof: MultiProof) -> Vec<MultiProof> {
+        if sequence < self.next_sequence {
+            return vec![proof];
+        }
```

Comment:
proofs are not guaranteed to arrive in order. this check doesn't make much sense to me

Reply:
yep good point, there are tests for this case but they were not properly defined, should be fixed now, ptal